bee-network/bee-autopeering/src/packet.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! IOTA network packets.
+
+use crate::{hash, peer::peer_id::PeerId, proto};
+
+use base64 as bs64;
+use crypto::{
+    hashes::sha::SHA256_LEN,
+    signatures::ed25519::{PublicKey, Signature},
+};
+use num_derive::FromPrimitive;
+use prost::{bytes::BytesMut, DecodeError, EncodeError, Message};
+
+use std::{fmt, net::SocketAddr, ops::Range};
+
+// From `hive.go` docs:
+// * specifies the maximum allowed size of packets;
+// * packets larger than this will be cut and thus treated as invalid;
+pub(crate) const MAX_PACKET_SIZE: usize = 1280;
+
+pub(crate) const DISCOVERY_MSG_TYPE_MIN: u32 = 10;
+pub(crate) const DISCOVERY_MSG_TYPE_RANGE: Range<u32> = DISCOVERY_MSG_TYPE_MIN..(DISCOVERY_MSG_TYPE_MIN + 4);
+pub(crate) const PEERING_MSG_TYPE_MIN: u32 = 20;
+pub(crate) const PEERING_MSG_TYPE_RANGE: Range<u32> = PEERING_MSG_TYPE_MIN..(PEERING_MSG_TYPE_MIN + 3);
+
+pub(crate) fn msg_hash(msg_type: MessageType, msg_data: &[u8]) -> [u8; SHA256_LEN] {
+    let mut bytes = vec![0u8; msg_data.len() + 1];
+    let msg_type = msg_type as u32;
+    if msg_type > 0xFF {
+        panic!("invalid message type");
+    }
+    bytes[0] = msg_type as u8;
+    bytes[1..].copy_from_slice(msg_data);
```

Comment:
It's a bit sad that crypto.rs doesn't expose the Digest features of its sha module, there is a lot of unnecessary allocating/copying because of that :(

Reply:
It does publicly re-export the symbols though https://github.com/iotaledger/crypto.rs/blob/dev/src/hashes/sha.rs#L9
So we should be able to use the `Digest` features. I would say it's worth to remove all the allocations and copies

Reply:
Refactored. We could even reuse the Sha256 instance so we don't have to initialize it each time, but let's keep those further optimizations for future releases.